### PR TITLE
M-FAC BERT Integration

### PR DESCRIPTION
**Pull request by KSGulin**
Integrated MFAC Into BERT
Updated flow to use small block version in all reasonable cases

### CODE REVIEW

File: src/sparseml/pytorch/optim/modifier_pruning.py [hunk 1]
```diff
@@ -1376,6 +1379,8 @@ class MFACPruningModifier(GMPruningModifier):
         channels, or a SparsityMaskCreator object. default is 'unstructured'
     :param global_sparsity: set True to enable global pruning. if False, pruning will
         be layer-wise. Default is False
+    :param use_gradient_buffering: Optional bool to use gradient buffering instead of
+    grad sampling. By default, grad sampling is always used when available
```
Comment (bfineran) [inline src/sparseml/pytorch/optim/modifier_pruning.py:1383]: nit: indent

File: src/sparseml/pytorch/optim/modifier_pruning.py [hunk 2]
```diff
@@ -1399,6 +1404,7 @@ def __init__(
         log_types: Union[str, List[str]] = ALL_TOKEN,
         mask_type: Union[str, List[int], PruningMaskCreator] = "unstructured",
         global_sparsity: bool = False,
+        use_gradient_buffering: Optional[bool] = None,
```
Comment (bfineran) [inline src/sparseml/pytorch/optim/modifier_pruning.py:1407]: need to add a `ModifierProp` for `use_gradient_buffering`

File: src/sparseml/pytorch/optim/modifier_pruning.py [hunk 3]
```diff
@@ -1511,8 +1529,10 @@ def _collect_grad_samples(
             self._applied_sparsity or 0.0
         )
 
+        _LOGGER.debug("Starting to collect {num_grads} grads with GradSampler")
```
Comment (bfineran) [inline src/sparseml/pytorch/optim/modifier_pruning.py:1532]:
```suggestion
        _LOGGER.debug(f"Starting to collect {num_grads} grads with GradSampler")
```

File: src/sparseml/pytorch/utils/mfac_helpers.py [hunk 4]
```diff
@@ -275,54 +280,19 @@ def __init__(self, grads, block_size, damp=1e-5, devices=None):
...
         self._fisher_inv_blocks = []
-        lock = threading.Lock()
-
-        # run block computations in parallel across devices
-        threads = [None] * len(self._devices)
-        thread_fisher_inv_blocks = [None] * len(self._devices)
 
-        def _compute_fisher_inv_block(block_start_idx, thread_idx_):
+        _LOGGER.debug("Starting FisherInverseFastBlock")
+        for block_start_idx in range(0, grads.shape[1], self._block_size):
```
Comment (bfineran) [inline src/sparseml/pytorch/utils/mfac_helpers.py:285]: is there any way to take advantage of multiple gpus here without using an external api?

File: src/sparseml/pytorch/utils/mfac_helpers.py [hunk 5]
```diff
@@ -622,25 +592,107 @@ def __init__(
...
+                )
+                prev_free_memory = free_device_memory
+                _LOGGER.debug(f"Free memory on {device} - {free_device_memory}")
+                if free_device_memory < prev_free_memory:
+                    _LOGGER.debug(
+                        f"WARNING - GPU memory not cleanly freed. Found {prev_free_memory - free_device_memory}"
+                        f"less bytes / {prev_free_memory - free_device_memory/BYTES_IN_GIB}"
+                        f"since the last iteration"
+                    )
+                self._num_blocks_per_device_call.append(
```
Comment (bfineran) [inline src/sparseml/pytorch/utils/mfac_helpers.py:624]: can you add a comment on what this does?

File: src/sparseml/pytorch/utils/mfac_helpers.py [hunk 6]
```diff
@@ -17,6 +17,7 @@
 pruning.
```
Comment (bfineran) [inline src/sparseml/pytorch/utils/mfac_helpers.py:17]: as we're getting more logic changes within the HInv classes, it would be great for us to add some testing that would compare results of HInv calcs across the implementations for the same matrix. feel free to sync on this if needed

File: src/sparseml/transformers/sparsification/trainer.py [hunk 7]
```diff
@@ -119,6 +119,10 @@ def __init__(
         self.callback_disable_fp16 = DisableHalfPrecisionCallback(self)
         self.callback_handler.add_callback(self.callback_disable_fp16)
 
+        self.grad_sampler = GradSampler(
```
Comment (bfineran) [inline src/sparseml/transformers/sparsification/trainer.py:122]: thinking about some changes we can make to the grad sampler to make the requirements for the dataloader a little less strict. let's sync on this tomorrow. similar to the changes requested in the dataloader here: https://github.com/neuralmagic/sparseml/pull/541

**Review by bfineran — COMMENTED** [review]
great work diving into this

File: src/sparseml/pytorch/utils/mfac_helpers.py (hunk 4 above)
Comment (KSGulin) [reply src/sparseml/pytorch/utils/mfac_helpers.py:287]: In my research I haven't been able to find another way. It seems in general there are two options - either code your own interactions with the CUDA kernel or rely on ones created by others (e.g. DPP in PyTorch). But I'd be curious to hear if anyone else on the team has had experience with this.

**Review by spacemanidol — APPROVED** [review]

**Review by bfineran — APPROVED** [review]
